src/dipdup/dipdup.py

```diff
@@ -268,38 +266,66 @@ async def _on_big_maps(self, datasource: TzktDatasource, big_maps: Tuple[BigMapD
 
     async def _on_rollback(self, datasource: TzktDatasource, from_level: int, to_level: int) -> None:
         """Perform a single level rollback when possible, otherwise call `on_rollback` hook"""
-        self._logger.warning('Datasource `%s` rolled back: %s -> %s', datasource.name, from_level, to_level)
+        if from_level <= to_level:
+            raise RuntimeError(f'Attempt to rollback forward: {from_level} -> {to_level}')
+
+        self._logger.warning('Datasource `%s` has rolled back: %s -> %s', datasource.name, from_level, to_level)
         if Metrics.enabled:
             Metrics.set_datasource_rollback(datasource.name)
 
-        # NOTE: Zero difference between levels means we received no operations/big_maps on this level and thus channel level hasn't changed
-        zero_level_rollback = from_level - to_level == 0
-        single_level_rollback = from_level - to_level == 1
-
-        if zero_level_rollback:
-            self._logger.info('Zero level rollback, ignoring')
-
-        elif single_level_rollback:
-            # NOTE: Notify all indexes which use rolled back datasource to drop duplicated operations from the next block
-            self._logger.info('Checking if single level rollback is possible')
-            matching_indexes = tuple(i for i in self._indexes.values() if i.datasource == datasource)
-            matching_operation_indexes = tuple(i for i in matching_indexes if isinstance(i, OperationIndex))
-            self._logger.info(
-                'Indexes: %s total, %s matching, %s support single level rollback',
-                len(self._indexes),
-                len(matching_indexes),
-                len(matching_operation_indexes),
-            )
+        # NOTE: Choose action for each index
+        ignored_indexes: Set[str] = set()
+        single_level_indexes: Set[str] = set()
+        unprocessed_indexes: Set[str] = set()
+
+        for index_name, index in self._indexes.items():
+            index_level = index.state.level
+
+            if index.datasource != datasource:
```

Comment:
What is the case?

Reply:
Multiple datasources for different networks in config. One can roll back while others not.